src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs

```diff
@@ -144,8 +147,16 @@ protected override void Dispose(bool disposing)
                 SafeSslHandle sslContext = _sslContext;
                 if (null != sslContext)
                 {
-                    _inputBuffer.Dispose();
-                    _outputBuffer.Dispose();
+                    if (Interlocked.Exchange(ref _pendingInput, 2) == 0)
+                    {
+                        _inputBuffer.Dispose();
+                    }
+
+                    if (Interlocked.Exchange(ref _pendingOutput, 2) == 0)
+                    {
+                        _outputBuffer.Dispose();
+                    }
```

Comment:
Would it make sense to wrap ArrayBuffer in a ReservableArrayBuffer, or something like that?  Then it would have methods like Rent/Return/Dispose, which would internally do the right synchronization on a ref count or such field.